solidity/contracts/middleware/InterchainAccountRouter.sol

```diff
@@ -813,18 +856,48 @@ contract InterchainAccountRouter is Router {
             );
     }
 
-    /// @dev The calls represented by the commitment can only be executed once.
-    function revealAndExecute(
-        CallLib.Call[] calldata _calls,
-        bytes32 _salt
-    ) external payable {
-        bytes32 _givenCommitment = keccak256(abi.encode(_salt, _calls));
-        OwnableMulticall ica = verifiedCommitments[_givenCommitment];
+    function callRemoteCommitReveal(
+        uint32 _destination,
+        bytes32 _commitment,
+        uint _gasLimit
+    ) public payable returns (bytes32 _commitmentMsgId, bytes32 _revealMsgId) {
+        bytes32 _router = routers(_destination);
+        bytes32 _ism = isms[_destination];
 
-        require(address(payable(ica)) != address(0), "Invalid Reveal");
+        bytes memory hookMetadata = StandardHookMetadata.overrideGasLimit(
+            _gasLimit
+        );
+
+        return
+            callRemoteCommitReveal(
+                _destination,
+                _router,
+                _ism,
+                bytes32(0),
+                hookMetadata,
+                hook,
+                InterchainAccountMessage.EMPTY_SALT,
+                _commitment
+            );
+    }
 
-        delete verifiedCommitments[_givenCommitment];
-        ica.multicall{value: msg.value}(_calls);
+    /// @dev The calls represented by the commitment can only be executed once per commitment,
+    /// though you can submit the same commitment again after the calls have been executed.
+    function revealAndExecute(
+        CallLib.Call[] calldata calls,
+        bytes32 salt,
+        OwnableMulticall ica
+    ) external payable returns (bytes32 executedCommitment) {
+        // If there is no active commitment, do nothing.
+        bytes32 icaCommitment = ica.commitment();
+        if (icaCommitment == bytes32(0)) return bytes32(0);
+
+        bytes32 revealedHash = keccak256(abi.encode(calls, salt, ica));
+        require(icaCommitment == revealedHash, "Invalid Reveal");
+
+        ica.deleteCommitment();
+        ica.multicall{value: msg.value}(calls);
```

Comment:
I think it would be far more efficient if this all happened inside of a single call to the ICA

Reply:
Updated